connections/muxer-sel-in-sec-handshake.md

```diff
@@ -0,0 +1,220 @@
+   be accomplished within the TLS handshake, without adding network
+   round-trips, and allows the server to associate a different
+   certificate with each application protocol, if desired.
+
+For the purpose of muxer selection, the types of muxers are coded as protocol
+names in the form of a list of strings, and inserted in the ALPN "NextProtos"
+field. An example list as following:
+
+    ["yamux/1.0.0", "mplux", "libp2p"]
+
+The NextProtos list is ordered by preference, with the most prefered muxer at
+the beginning. The "libp2p" protocol code MUST always be the last item in the
+ALPN NextProtos list.
+
+The server chooses the supported protocol by going through its prefered
+protocol list and searchs if the protocol is supported by the client too. if no
+mutually supported protcol is found the TLS handshake will fail.
+
+If the selected NextProto is "libp2p" then the muxer selection process returns
+an empty result, and the multistream-selection protocol MUST be run to negotiate
+the muxer.
+
+(TBD: in some cases early abortion is desireable)
+
+
+### Muxer selection over Noise
+
+The libp2p Noise implementation allows the Noise handshake process to carry
+early data. [Noise-Early-Data] is carried in the second and third message of
+the handshake pattern:
+
+    XX:
+    -> e
+    <- e, ee, s, es
+    -> s, se
+
+At the end of the handshake pattern, both the client and server have received
+the peer's early data. The Noise protocol does not perform the protocol
+selection as TLS does, rather it just delivers the early data to handshaking
+peers.
+
+The muxer selection logic runs out of the Noise handshake process, relying on
+the early data exchanged during the handshake. The early data is delivered in
+the form of a byte string. The supported muxers are passed in space separated
+string codes. An example early data string:
+
+    "yamux/1.0.0 mplux"
+
+The byte string is ordered by preference, with the most prefered muxer at the
+beginning.
+
+After the Noise handshake, the client and server run the muxer selection
+process with the same logic. Each side will go through the server's early
+data from most prefered to lest prefered muxer, and if the muxer is in the
+client's early data list, that muxer is selected. The process guarantees that
+both the client and server reaches at the same conclusion of muxer slection.
+
+If the muxer selection process does not find any mutually supported muxer, for
+example, in the case that one early data string is empty, then an empty muxer
+selection result is returned, and multistream-selection MUST be performed.
+
+(TBD: in some cases early abortion is desireable)
+
+## Cross version support
+
+The improved muxer selection approach MUST be interoperable with pervious
+libp2p versions which do not support this improved approach.
+
+### TLS case
+
+In the current version of libp2p, the "NextProtos" field is populated with a
+key "libp2p". By appending the key of "libp2p" to the end of the supported
```

Comment:
That's correct, I double checked the go implementation, Here is what it says: 
        // NextProtos is a list of supported application level protocols, in
	// order of preference. If both peers support ALPN, the selected
	// protocol will be one from this list, and the connection will fail
	// if there is no mutually supported protocol. If NextProtos is empty
	// or the peer doesn't support ALPN, the connection will succeed and
	// ConnectionState.NegotiatedProtocol will be empty.

The connection fails if there is not mutually supported ALPN, and can succeed if one of the ALPN fields is empty.  So we don't have to append "libp2p".

Reply:
We should probably document this. I don't think we have anything about this in this repo.

Reply:
>  If both peers support ALPN, the selected
> protocol will be one from this list, and the connection will fail
> if there is no mutually supported protocol.

I'm surprised by this. I _thought_ that the Go standard library didn't enforce this. RFC 7301 doesn't require peers to  fail the handshake if ALPN negotiation fails, which is why we had to require that behavior in QUIC (see https://datatracker.ietf.org/doc/html/rfc9001#section-8.1).

As @MarcoPolo said, I agree that it would be valuable to be very precise about what we expect in various case (ALPN offered with muxers, APLN offered with only libp2p, no / empty ALPN offered), and to have tests for this behavior in our implementation.

Reply:
Of course we will create test cases for this. It seems there are some differences between the RFC and the go lib implementation documentation. We need to flush out what is the actual behavior.